File: src/main/scala/resonant/lib/content/module/TileBlock.java
package resonant.lib.content.module;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.WeakHashMap;

import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Icon;
import net.minecraft.util.MathHelper;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeDirection;
import resonant.api.IRotatable;
import resonant.lib.prefab.item.ItemBlockTooltip;
import resonant.lib.prefab.vector.Cuboid;
import resonant.lib.utility.LanguageUtility;
import resonant.lib.utility.WrenchUtility;
import universalelectricity.api.vector.Vector2;
import universalelectricity.api.vector.Vector3;
import universalelectricity.api.vector.VectorWorld;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/** All blocks inherit this class.
 * <p/>
 * Note that a lot of the variables will not exist except on the primary instance of the TileBlock,
 * hosted in BlockDummy.
 * 
 * @author Calclavia */
public abstract class TileBlock extends TileEntity
{
    /** The unique string ID of this block. */
    public final String name;
    public final Material material;
    public Class<? extends ItemBlock> itemBlock = ItemBlockTooltip.class;
    public CreativeTabs creativeTab = null;
    public Cuboid bounds = Cuboid.full();
    public BlockDummy block;
    public float blockHardness = 1;
    public float blockResistance = 1;
    public boolean canProvidePower = false;
    public boolean tickRandomly = false;
    public boolean normalRender = true;
    public boolean forceStandardRender = false;
    public boolean customItemRender = false;
    public boolean isOpaqueCube = true;

    /** Temporary */
    public IBlockAccess access;
    /** Rendering */
    protected String textureName;
    protected String domain;
    /** Rotation */
    protected byte rotationMask = Byte.parseByte("111100", 2);
    protected boolean isFlipPlacement = false;

    public TileBlock(String newName, Material newMaterial)
    {
        name = newName;
        material = newMaterial;
        textureName = name;
    }

    public TileBlock(Material newMaterial)
    {
        name = LanguageUtility.decapitalizeFirst(getClass().getSimpleName().replaceFirst("Tile", ""));
        material = newMaterial;
        textureName = name;
    }

    public static Vector2 getClickedFace(byte hitSide, float hitX, float hitY, float hitZ)
    {
        switch (hitSide)
        {
            case 0:
                return new Vector2(1 - hitX, hitZ);
            case 1:
                return new Vector2(hitX, hitZ);
            case 2:
                return new Vector2(1 - hitX, 1 - hitY);
            case 3:
                return new Vector2(hitX, 1 - hitY);
            case 4:
                return new Vector2(hitZ, 1 - hitY);
            case 5:
                return new Vector2(1 - hitZ, 1 - hitY);
            default:
                return new Vector2(0.5, 0.5);
        }
    }

    /** Called after the block is registred. Use this to add recipes. */
    public void onInstantiate()
    {

    }

    public World world()
    {
        return worldObj;
    }

    public IBlockAccess access()
    {
        if (world() != null)
        {
            return world();
        }

        return access;
    }

    public int x()
    {
        assert world() != null : "TileBlock [" + getClass().getSimpleName() + "] attempted to access invalid method.";
        return xCoord;
    }

    public int y()
    {
        assert world() != null : "TileBlock [" + getClass().getSimpleName() + "] attempted to access invalid method.";
        return yCoord;
    }

    public int z()
    {
        assert world() != null : "TileBlock [" + getClass().getSimpleName() + "] attempted to access invalid method.";
        return zCoord;
    }

    public VectorWorld position()
    {
        assert world() != null : "TileBlock [" + getClass().getSimpleName() + "] attempted to access invalid method.";
        return new VectorWorld(this);
    }

    protected VectorWorld center()
    {
        assert world() != null : "TileBlock [" + getClass().getSimpleName() + "] attempted to access invalid method.";
        return (VectorWorld) position().translate(0.5);
    }

    @Override
    public Block getBlockType()
    {
        if (access() != null)
        {
            Block b = Block.blocksList[access().getBlockId(this.xCoord, this.yCoord, this.zCoord)];
            if (b == null)
            {
                return block;
            }
            return b;
        }

        return block;
    }

    /** @return Return "this" if the block requires a TileEntity. */
    public TileBlock tile()
    {
        return null;
    }

    /** @return The containing block. */
    public Block block()
    {
        return Block.blocksList[blockID()];
    }

    public int blockID()
    {
        if (access() != null)
        {
            return access().getBlockId(x(), y(), z());
        }

        return block.blockID;
    }

    public int metadata()
    {
        return access().getBlockMetadata(x(), y(), z());
    }

    /** Drops */
    public ArrayList<ItemStack> getDrops(int metadata, int fortune)
    {
        ArrayList<ItemStack> drops = new ArrayList<ItemStack>();
        if (getBlockType() != null)
        {
            drops.add(new ItemStack(getBlockType(), quantityDropped(metadata, fortune), metadataDropped(metadata, fortune)));
        }
        return drops;
    }

    public int quantityDropped(int meta, int fortune)
    {
        return 1;
    }

    public int metadataDropped(int meta, int fortune)
    {
        return 0;
    }

    public boolean isControlDown(EntityPlayer player)
    {
        try
        {
            Class ckm = Class.forName("codechicken.multipart.ControlKeyModifer");
            Method m = ckm.getMethod("isControlDown", EntityPlayer.class);
            return (Boolean) m.invoke(null, player);
        }
        catch (Exception e)
        {

        }
        return false;
    }

    public void getSubBlocks(int id, CreativeTabs creativeTab, List list)
    {
        list.add(new ItemStack(id, 1, 0));
    }

    public ItemStack getPickBlock(MovingObjectPosition target)
    {
        return new ItemStack(getBlockType(), 1, metadataDropped(metadata(), 0));
    }

    public int getLightValue(IBlockAccess access)
    {
        return Block.lightValue[access.getBlockId(x(), y(), z())];
    }

    /** Block events */
    public void click(EntityPlayer player)
    {

    }

    public boolean activate(EntityPlayer player, int side, Vector3 hit)
    {
        /** Check if the player is holding a wrench. If so, call the wrench event. */
        if (WrenchUtility.isUsableWrench(player, player.inventory.getCurrentItem(), x(), y(), z()))
        {
            if (configure(player, side, hit))
            {
                WrenchUtility.damageWrench(player, player.inventory.getCurrentItem(), x(), y(), z());
                return true;
            }

            return false;
        }

        return use(player, side, hit);
    }

    /** Called when the block is clicked by a player */
    protected boolean use(EntityPlayer player, int side, Vector3 hit)
    {
        return false;
    }

    /** Called when the block is clicked with a wrench */
    protected boolean configure(EntityPlayer player, int side, Vector3 hit)
    {
        return tryRotate(side, hit);
    }

    /** Rotatable Block */
    protected boolean tryRotate(int side, Vector3 hit)
    {
        if (this instanceof IRotatable)
        {
            byte result = getSideToRotate((byte) side, hit.x, hit.y, hit.z);

            if (result != -1)
            {
                setDirection(ForgeDirection.getOrientation(result));
                return true;
            }
        }

        return false;
    }

    /** @author Based of Greg (GregTech) */
    public byte getSideToRotate(byte hitSide, double hitX, double hitY, double hitZ)
    {
        byte tBack = (byte) (hitSide ^ 1);
        switch (hitSide)
        {
            case 0:
            case 1:
                if (hitX < 0.25)
                {
                    if (hitZ < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitZ > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(4))
                    {
                        return 4;
                    }
                }
                if (hitX > 0.75)
                {
                    if (hitZ < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitZ > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(5))
                    {
                        return 5;
                    }
                }
                if (hitZ < 0.25)
                {
                    if (canRotate(2))
                    {
                        return 2;
                    }
                }
                if (hitZ > 0.75)
                {
                    if (canRotate(3))
                    {
                        return 3;
                    }
                }
                if (canRotate(hitSide))
                {
                    return hitSide;
                }
            case 2:
            case 3:
                if (hitX < 0.25)
                {
                    if (hitY < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitY > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(4))
                    {
                        return 4;
                    }
                }
                if (hitX > 0.75)
                {
                    if (hitY < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitY > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(5))
                    {
                        return 5;
                    }
                }
                if (hitY < 0.25)
                {
                    if (canRotate(0))
                    {
                        return 0;
                    }
                }
                if (hitY > 0.75)
                {
                    if (canRotate(1))
                    {
                        return 1;
                    }
                }
                if (canRotate(hitSide))
                {
                    return hitSide;
                }
            case 4:
            case 5:
                if (hitZ < 0.25)
                {
                    if (hitY < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitY > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(2))
                    {
                        return 2;
                    }
                }
                if (hitZ > 0.75)
                {
                    if (hitY < 0.25)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (hitY > 0.75)
                    {
                        if (canRotate(tBack))
                        {
                            return tBack;
                        }
                    }
                    if (canRotate(3))
                    {
                        return 3;
                    }
                }
                if (hitY < 0.25)
                {
                    if (canRotate(0))
                    {
                        return 0;
                    }
                }
                if (hitY > 0.75)
                {
                    if (canRotate(1))
                    {
                        return 1;
                    }
                }
                if (canRotate(hitSide))
                {
                    return hitSide;
                }
        }
        return -1;
    }

    public ForgeDirection determineOrientation(EntityLivingBase entityLiving)
    {
        if (MathHelper.abs((float) entityLiving.posX - x()) < 2.0F && MathHelper.abs((float) entityLiving.posZ - z()) < 2.0F)
        {
            double d0 = entityLiving.posY + 1.82D - entityLiving.yOffset;

            if (canRotate(1) && d0 - y() > 2.0D)
            {
                return ForgeDirection.UP;
            }

            if (canRotate(0) && y() - d0 > 0.0D)
            {
                return ForgeDirection.DOWN;
            }
        }

        int playerSide = MathHelper.floor_double(entityLiving.rotationYaw * 4.0F / 360.0F + 0.5D) & 3;
        int returnSide = (playerSide == 0 && canRotate(2)) ? 2 : ((playerSide == 1 && canRotate(5)) ? 5 : ((playerSide == 2 && canRotate(3)) ? 3 : ((playerSide == 3 && canRotate(4)) ? 4 : 0)));

        if (isFlipPlacement)
        {
            return ForgeDirection.getOrientation(returnSide).getOpposite();
        }

        return ForgeDirection.getOrientation(returnSide);
    }

    public boolean canRotate(int ord)
    {
        return (rotationMask & (1 << ord)) != 0;
    }

    public ForgeDirection getDirection()
    {
        return ForgeDirection.getOrientation(metadata());
    }

    public void setDirection(ForgeDirection direction)
    {
        world().setBlockMetadataWithNotify(x(), y(), z(), direction.ordinal(), 3);
    }

    /** Block events */
    protected void onAdded()
    {
        onWorldJoin();
    }

    public void onPlaced(EntityLivingBase entityLiving, ItemStack itemStack)
    {
        if (this instanceof IRotatable)
        {
            ((IRotatable) this).setDirection(determineOrientation(entityLiving));
        }
    }

    public void onRemove(int par5, int par6)
    {
        onWorldSeparate();
    }

    protected void onWorldJoin()
    {
    }

    protected void onWorldSeparate()
    {
    }

    protected void onNeighborChanged()
    {
    }

    protected void notifyChange()
    {
        world().notifyBlocksOfNeighborChange(x(), y(), z(), blockID());
    }

    protected void markRender()
    {
        world().markBlockForRenderUpdate(x(), y(), z());
    }

    protected void markUpdate()
    {
        world().markBlockForUpdate(x(), y(), z());
    }

    protected void updateLight()
    {
        world().updateAllLightTypes(x(), y(), z());
    }

    protected void scheduelTick(int delay)
    {
        world().scheduleBlockUpdate(x(), y(), z(), blockID(), delay);
    }

    /** Called when an entity collides with this block. */
    public void collide(Entity entity)
    {

    }

    /** Collision Note that all bounds done in the the tile is relative to the tile's position. */
    public Iterable<Cuboid> getCollisionBoxes(Cuboid intersect, Entity entity)
    {
        List<Cuboid> boxes = new ArrayList<Cuboid>();

        for (Cuboid cuboid : getCollisionBoxes())
        {
            if (intersect != null && cuboid.intersects(intersect))
            {
                boxes.add(cuboid);
            }
        }

        return boxes;
    }

    public Iterable<Cuboid> getCollisionBoxes()
    {
        return Arrays.asList(new Cuboid[] { bounds });
    }

    public Cuboid getSelectBounds()
    {
        return bounds;
    }

    public Cuboid getCollisionBounds()
    {
        return bounds;
    }

    @SideOnly(Side.CLIENT)
    public final TileRender getRenderer()
    {
        // TODO: Be careful if this might cause memory issues.
        if (!RenderInfo.renderer.containsKey(this))
        {
            RenderInfo.renderer.put(this, newRenderer());
        }

        return RenderInfo.renderer.get(this);
    }

    @SideOnly(Side.CLIENT)
    protected TileRender newRenderer()
    {
        return null;
    }

    /** Called in the world. */
    @SideOnly(Side.CLIENT)
    public Icon getIcon(IBlockAccess access, int side)
    {
        return getIcon(side, access.getBlockMetadata(x(), y(), z()));
    }

    /** Called either by an item, or in a world. */
    @SideOnly(Side.CLIENT)
    public Icon getIcon(int side, int meta)
    {
        return getIcon();
    }

    public Icon getIcon()
    {
        return RenderInfo.icon.get(getTextureName());
    }

    @SideOnly(Side.CLIENT)
    public void registerIcons(IconRegister iconRegister)
    {
        RenderInfo.icon.put(getTextureName(), iconRegister.registerIcon(getTextureName()));
    }

    @SideOnly(Side.CLIENT)
    protected String getTextureName()
    {
        return textureName == null ? "MISSING_ICON_TILE_" + getBlockType().blockID + "_" + name : block.dummyTile.domain + textureName;
    }

    public boolean shouldSideBeRendered(IBlockAccess access, int x, int y, int z, int side)
    {
        return side == 0 && this.bounds.min.y > 0.0D ? true : (side == 1 && this.bounds.max.y < 1.0D ? true : (side == 2 && this.bounds.min.z > 0.0D ? true : (side == 3 && this.bounds.max.z < 1.0D ? true : (side == 4 && this.bounds.min.x > 0.0D ? true : (side == 5 && this.bounds.max.x < 1.0D ? true : !access.isBlockOpaqueCube(x, y, z))))));
    }

    public void onFillRain()
    {

    }

    public boolean isIndirectlyPowered()
    {
        return world().isBlockIndirectlyGettingPowered(x(), y(), z());
    }

    public int getStrongestIndirectPower()
    {
        return world().getStrongestIndirectPower(x(), y(), z());
    }

    public int getWeakRedstonePower(IBlockAccess access, int side)
    {
        return getStrongRedstonePower(access, side);
    }

    public int getStrongRedstonePower(IBlockAccess access, int side)
    {
        return 0;
    }

    public boolean isSolid(IBlockAccess access, int side)
    {
        return material.isSolid();
    }

    public int getRenderBlockPass()
    {
        return 0;
    }

    public int tickRate(World world)
    {
        return 20;
    }

    public interface IComparatorInputOverride
    {
        public int getComparatorInputOverride(int side);
    }

    /** Rendering */
    @SideOnly(Side.CLIENT)
    public static class RenderInfo
    {
        @SideOnly(Side.CLIENT)
        private static final WeakHashMap<TileBlock, TileRender> renderer = new WeakHashMap<TileBlock, TileRender>();

        @SideOnly(Side.CLIENT)
        private static final HashMap<String, Icon> icon = new HashMap<String, Icon>();
    }

}

File: src/main/scala/resonant/lib/content/module/TileBlock.scala
package resonant.lib.content.module

import cpw.mods.fml.relauncher.Side
import cpw.mods.fml.relauncher.SideOnly
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.Item
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.IIcon
import net.minecraft.util.MathHelper
import net.minecraft.util.MovingObjectPosition
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection
import resonant.api.IRotatable
import resonant.lib.prefab.item.ItemBlockTooltip
import resonant.lib.utility.LanguageUtility
import resonant.lib.utility.WrenchUtility
import universalelectricity.core.transform.region.Cuboid
import universalelectricity.core.transform.vector.Vector2
import universalelectricity.core.transform.vector.Vector3
import universalelectricity.core.transform.vector.VectorWorld
import java.lang.reflect.Method
import java.util._
import resonant.lib.content.module.TileBlock.RenderInfo
import java.lang.Byte._
import scala.collection.immutable
import resonant.lib.content.prefab.TraitRotatable

/**
 * All blocks inherit this class.
 * <p/>
 * Note that a lot of the variables will not exist except on the primary instance of the TileBlock,
 * hosted in BlockDummy.
 *
 * @author Calclavia
 */
object TileBlock
{
  def getClickedFace(hitSide: Byte, hitX: Float, hitY: Float, hitZ: Float): Vector2 =
  {
    hitSide match
    {
      case 0 =>
        return new Vector2(1 - hitX, hitZ)
      case 1 =>
        return new Vector2(hitX, hitZ)
      case 2 =>
        return new Vector2(1 - hitX, 1 - hitY)
      case 3 =>
        return new Vector2(hitX, 1 - hitY)
      case 4 =>
        return new Vector2(hitZ, 1 - hitY)
      case 5 =>
        return new Vector2(1 - hitZ, 1 - hitY)
      case _ =>
        return new Vector2(0.5, 0.5)
    }
  }

  abstract trait IComparatorInputOverride
  {
    def getComparatorInputOverride(side: Int): Int
  }

  /**
   * Rendering
   */
  @SideOnly(Side.CLIENT) object RenderInfo
  {
    @SideOnly(Side.CLIENT) final val renderer: WeakHashMap[TileBlock, TileRender] = new WeakHashMap[TileBlock, TileRender]
    @SideOnly(Side.CLIENT) final val icon: HashMap[String, IIcon] = new HashMap[String, IIcon]
  }

}

abstract class TileBlock(val name: String, val material: Material) extends TileEntity
{
  /**
   * The unique string ID of this block.
   */
  var itemBlock: Class[_ <: ItemBlock] = classOf[ItemBlockTooltip]
  var creativeTab: CreativeTabs = null
  var bounds: Cuboid = Cuboid.full
  var block: BlockDummy = null
  var blockHardness: Float = 1
  var blockResistance: Float = 1
  var canProvidePower: Boolean = false
  var tickRandomly: Boolean = false
  var normalRender: Boolean = true
  var forceStandardRender: Boolean = false
  var customItemRender: Boolean = false
  var isOpaqueCube: Boolean = true
  /**
   * Temporary
   */
  var _access: IBlockAccess = null
  /**
   * Rendering
   */
  var textureName: String = name
  var domain: String = null
  /**
   * Rotation
   */
  protected var rotationMask: Byte = parseByte("111100", 2)
  protected var isFlipPlacement: Boolean = false

  def this(newMaterial: Material) = this(LanguageUtility.decapitalizeFirst(getClass.getSimpleName.replaceFirst("Tile", "")), newMaterial)


  /**
   * Called after the block is registred. Use this to add recipes.
   */
  def onInstantiate
  {
  }

  def world: World =
  {
    return worldObj
  }

  def world(world: World)
  {
    this.worldObj = world
  }

  def access: IBlockAccess =
  {
    if (world != null)
    {
      return world
    }

    return _access
  }

  def x: Int =
  {
    assert(world != null, "TileBlock [" + getClass.getSimpleName + "] attempted to access invalid method.")
    return xCoord
  }

  def y: Int =
  {
    assert(world != null, "TileBlock [" + getClass.getSimpleName + "] attempted to access invalid method.")
    return yCoord
  }

  def z: Int =
  {
    assert(world != null, "TileBlock [" + getClass.getSimpleName + "] attempted to access invalid method.")
    return zCoord
  }

  def position: VectorWorld =
  {
    assert(world != null, "TileBlock [" + getClass.getSimpleName + "] attempted to access invalid method.")
    return new VectorWorld(this)
  }

  protected def center: VectorWorld =
  {
    assert(world != null, "TileBlock [" + getClass.getSimpleName + "] attempted to access invalid method.")
    return position.add(0.5).asInstanceOf[VectorWorld]
  }

  override def getBlockType: Block =
  {
    if (access != null)
    {
      val b: Block = access.getBlock(x, y, z)
      if (b == null)
      {
        return block
      }
      return b
    }
    return block
  }

  /**
   * @return Return "this" if the block requires a TileEntity.
   */
  def tile: TileBlock =
  {
    return null
  }

  def metadata: Int =
  {
    return access.getBlockMetadata(x, y, z)
  }

  /**
   * Drops
   */
  def getDrops(metadata: Int, fortune: Int): ArrayList[ItemStack] =
  {
    val drops: ArrayList[ItemStack] = new ArrayList[ItemStack]
    if (getBlockType != null)
    {
      drops.add(new ItemStack(getBlockType, quantityDropped(metadata, fortune), metadataDropped(metadata, fortune)))
    }
    return drops
  }

  def quantityDropped(meta: Int, fortune: Int): Int =
  {
    return 1
  }

  def metadataDropped(meta: Int, fortune: Int): Int =
  {
    return 0
  }

  def isControlDown(player: EntityPlayer): Boolean =
  {
    try
    {
      val ckm: Class[_] = Class.forName("codechicken.multipart.ControlKeyModifer")
      val m: Method = ckm.getMethod("isControlDown", classOf[EntityPlayer])
      return m.invoke(null, player).asInstanceOf[Boolean]
    }
    catch
      {
        case e: Exception =>
        {
        }
      }
    return false
  }

  def getSubBlocks(item: Item, creativeTabs: CreativeTabs, list: List[_])
  {
    def add[T](list: java.util.List[T], value: Any) = list.add(value.asInstanceOf[T])
    add(list, new ItemStack(item))
  }

  def getPickBlock(target: MovingObjectPosition): ItemStack =
  {
    return new ItemStack(getBlockType, 1, metadataDropped(metadata, 0))
  }

  def getLightValue(access: IBlockAccess): Int =
  {
    return block.getLightValue
  }

  /**
   * Block events
   */
  def click(player: EntityPlayer)
  {
  }

  def activate(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (WrenchUtility.isUsableWrench(player, player.inventory.getCurrentItem, x, y, z))
    {
      if (configure(player, side, hit))
      {
        WrenchUtility.damageWrench(player, player.inventory.getCurrentItem, x, y, z)
        return true
      }
      return false
    }
    return use(player, side, hit)
  }

  /**
   * Called when the block is clicked by a player
   */
  protected def use(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    return false
  }

  /**
   * Called when the block is clicked with a wrench
   */
  protected def configure(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (this.isInstanceOf[TraitRotatable])
      return this.asInstanceOf[TraitRotatable].rotate(side, hit)

    return false
  }



  /**
   * Block events
   */
  def onAdded
  {
    onWorldJoin
  }

  def onPlaced(entityLiving: EntityLivingBase, itemStack: ItemStack)
  {
    if (this.isInstanceOf[TraitRotatable])
    {
      this.asInstanceOf[TraitRotatable].setDirection(this.asInstanceOf[TraitRotatable].determineRotation(entityLiving))
    }
  }

  def onRemove(block: Block, par6: Int)
  {
    onWorldSeparate
  }

  def onWorldJoin
  {
  }

  def onWorldSeparate
  {
  }

  def onNeighborChanged(block: Block)
  {
  }

  def notifyChange
  {
    world.notifyBlocksOfNeighborChange(x, y, z, block)
  }

  protected def markRender
  {
    world.func_147479_m(x, y, z)
  }

  protected def markUpdate
  {
    world.markBlockForUpdate(x, y, z)
  }

  protected def updateLight
  {
    world.func_147451_t(x, y, z)
  }

  protected def scheduelTick(delay: Int)
  {
    world.scheduleBlockUpdate(x, y, z, block, delay)
  }

  /**
   * Called when an entity collides with this block.
   */
  def collide(entity: Entity)
  {
  }

  /**
   * Collision Note that all bounds done in the the tile is relative to the tile's position.
   */
  def getCollisionBoxes(intersect: Cuboid, entity: Entity): Iterable[Cuboid] =
  {
    val boxes: List[Cuboid] = new ArrayList[Cuboid]
    import scala.collection.JavaConversions._
    for (cuboid <- getCollisionBoxes)
    {
      if (intersect != null && cuboid.intersects(intersect))
      {
        boxes.add(cuboid)
      }
    }
    return boxes
  }

  def getCollisionBoxes: Iterable[Cuboid] =
  {
    return immutable.List[Cuboid](bounds)
  }

  def getSelectBounds: Cuboid =
  {
    return bounds
  }

  def getCollisionBounds: Cuboid =
  {
    return bounds
  }

  @SideOnly(Side.CLIENT) final def getRenderer: TileRender =
  {
    if (!RenderInfo.renderer.containsKey(this))
    {
      RenderInfo.renderer.put(this, newRenderer)
    }
    return RenderInfo.renderer.get(this)
  }

  @SideOnly(Side.CLIENT) protected def newRenderer: TileRender =
  {
    return null
  }

  /**
   * Called in the world.
   */
  @SideOnly(Side.CLIENT) def getIcon(access: IBlockAccess, side: Int): IIcon =
  {
    return getIcon(side, access.getBlockMetadata(x, y, z))
  }

  /**
   * Called either by an item, or in a world.
   */
  @SideOnly(Side.CLIENT) def getIcon(side: Int, meta: Int): IIcon =
  {
    return getIcon
  }

  def getIcon: IIcon =
  {
    return RenderInfo.icon.get(getTextureName)
  }

  @SideOnly(Side.CLIENT) def registerIcons(iconRegister: IIconRegister)
  {
    RenderInfo.icon.put(getTextureName, iconRegister.registerIcon(getTextureName))
  }

  @SideOnly(Side.CLIENT) protected def getTextureName: String =
  {
    return if (textureName == null) "MISSING_ICON_TILE_" + Block.getIdFromBlock(block) + "_" + name else block.dummyTile.domain + textureName
  }

  def shouldSideBeRendered(access: IBlockAccess, x: Int, y: Int, z: Int, side: Int): Boolean =
  {
    return if (side == 0 && this.bounds.min.y > 0.0D) true else (if (side == 1 && this.bounds.max.y < 1.0D) true else (if (side == 2 && this.bounds.min.z > 0.0D) true else (if (side == 3 && this.bounds.max.z < 1.0D) true else (if (side == 4 && this.bounds.min.x > 0.0D) true else (if (side == 5 && this.bounds.max.x < 1.0D) true else !access.getBlock(x, y, z).isOpaqueCube)))))
  }

  def onFillRain
  {
  }

  def isIndirectlyPowered: Boolean =
  {
    return world.isBlockIndirectlyGettingPowered(x, y, z)
  }

  def getStrongestIndirectPower: Int =
  {
    return world.getStrongestIndirectPower(x, y, z)
  }

  def getWeakRedstonePower(access: IBlockAccess, side: Int): Int =
  {
    return getStrongRedstonePower(access, side)
  }

  def getStrongRedstonePower(access: IBlockAccess, side: Int): Int =
  {
    return 0
  }

  def isSolid(access: IBlockAccess, side: Int): Boolean =
  {
    return material.isSolid
  }

  def getRenderBlockPass: Int =
  {
    return 0
  }

  def tickRate(world: World): Int =
  {
    return 20
  }

}